src/client.rs

```diff
@@ -250,21 +262,119 @@ impl<S: Read + Write> Client<S> {
-        self.stream.write_all(&raw)?;
-        self.stream.flush()?;
+        let mut stream = self.stream.lock().unwrap();
+        stream.write_all(&raw)?;
+        stream.flush()?;
+        drop(stream); // release the lock
 
         self.increment_calls();
 
-        let mut resp = loop {
-            let raw = self.recv()?;
-            let mut resp: serde_json::Value = serde_json::from_slice(&raw)?;
-
-            match resp["method"].take().as_str() {
-                Some(ref method) if method == &req.method => break resp,
-                Some(ref method) => self.handle_notification(method, resp["result"].take())?,
-                _ => break resp,
-            };
+        let mut raw_resp = String::new();
+        let mut resp = 'outer: loop {
+            // Try to take the lock on the reader. If we manage to do so, we'll become the reader
+            // thread until we get our reponse
+            match self.buf_reader.try_lock() {
+                Ok(mut reader) => {
+                    trace!("Thread for message {} became the reader thread", req.id);
+
+                    // Start a loop to wait for our message
+                    loop {
+                        raw_resp.clear();
+
+                        reader.read_line(&mut raw_resp)?;
+                        trace!("<== {}", raw_resp);
+
+                        let resp: serde_json::Value = serde_json::from_str(&raw_resp)?;
+
+                        /*match resp["method"].take().as_str() {
+                            Some(ref method) if method == &req.method => break 'outer resp,
+                            Some(ref method) => self.handle_notification(method, resp["result"].take())?,
+                            _ => break resp,
+                        };*/
+
+                        // Normally there is and id, but it's missing for spontaneous notifications
+                        // from the server
+                        let resp_id = resp["id"]
+                            .as_str()
+                            .and_then(|s| s.parse().ok())
+                            .or(resp["id"].as_u64().map(|i| i as usize));
+                        match resp_id {
+                            Some(resp_id) if resp_id == req.id => {
+                                // We have a valid id and it's exactly the one we were waiting for!
+                                trace!(
+                                    "Reader thread {} received a response for its request",
+                                    req.id
+                                );
+
+                                // Remove ourselves from the "waiting map"
+                                let mut map = self.waiting_map.write().unwrap();
+                                map.remove(&req.id);
+
+                                // If the map is not empty, we select a random thread to become the
+                                // new reader thread. We close its channel to "wake it up"
```

Comment:
maybe clearer to use an explicit enum as the value sent across the channel to receive data or be woke up, something like:
```
enum ValOrWake {
 Val(Value),
WakeUp,
}
```

Reply:
yeah that's a good idea!